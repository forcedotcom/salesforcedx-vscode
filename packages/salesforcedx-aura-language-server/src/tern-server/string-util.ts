const isAlphaNumberic = (code: number): boolean => {
    if (code > 47 && code < 58) {
        // numeric
        return true;
    }
    if ((code > 64 && code < 91) || (code > 96 && code < 123)) {
        // alpha
        return true;
    }
    if (code === 95 || code === 45) {
        // _ and -
        return true;
    }
    return false;
};

export const findWord = (str: string, offset: number): { start: number; end: number } => {
    let start = -1;
    let end = -1;

    let pos: number = offset;
    let c: number;

    while (pos >= 0) {
        c = str.charCodeAt(pos);
        if (c === 10 || c === 13) {
            // \n and \r
            break;
        } else if (!isAlphaNumberic(c)) {
            pos++;
            break;
        }
        --pos;
    }
    start = pos;

    pos = offset;
    const length = str.length;

    while (pos < length) {
        c = str.charCodeAt(pos);
        if (c === 10 || c === 13) {
            // \n and \r
            break;
        } else if (!isAlphaNumberic(c)) {
            break;
        }
        ++pos;
    }
    end = pos;

    if (start > -1 && end > -1) {
        return {
            start,
            end,
        };
    }
};

export const countPreviousCommas = (str: string, offset: number): number => {
    let commas = 0;
    let pos: number = offset;
    let c: number;
    while (pos >= 0) {
        c = str.charCodeAt(pos);
        if (c === 10 || c === 13 || c === 40) {
            // \n and \r and ()
            break;
        } else if (c === 44) {
            // ,
            commas++;
        }
        --pos;
    }
    return commas;
};

export const findPreviousLeftParan = (str: string, offset: number): number => {
    let start = -1;
    let pos: number = offset;
    let c: number;
    while (pos >= 0) {
        c = str.charCodeAt(pos);
        if (c === 10 || c === 13) {
            // \n and \r
            break;
        } else if (c === 40) {
            start = pos;
            break;
        }
        --pos;
    }
    return start;
};

export const findPreviousWord = (str: string, offset: number): { start: number; end: number } => {
    let start = -1;
    let end = -1;

    let seenWordBoundary = false;
    let boundaryOffset = 0;

    let pos: number = offset;
    let c: number;

    while (pos >= 0) {
        c = str.charCodeAt(pos);
        if (c === 10 || c === 13) {
            // \n and \r
            break;
        } else if ((c === 46 || c === 40) && !seenWordBoundary) {
            // . and (
            seenWordBoundary = true;
            boundaryOffset = offset - pos + 1;
        } else if (!isAlphaNumberic(c)) {
            pos++;
            break;
        }
        --pos;
    }
    if (!seenWordBoundary) {
        return {
            start: offset,
            end: offset,
        };
    }
    start = pos;

    pos = offset;
    const length: number = str.length;

    while (pos < length) {
        c = str.charCodeAt(pos);
        if (c === 10 || c === 13) {
            // \n and \r
            break;
        } else if (!isAlphaNumberic(c)) {
            break;
        }
        ++pos;
    }
    end = pos;

    if (start > -1 && end > -1) {
        return { start, end: end - boundaryOffset };
    }
};
