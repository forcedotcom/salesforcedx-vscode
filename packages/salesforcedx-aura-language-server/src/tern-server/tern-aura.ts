/* eslint-disable @typescript-eslint/no-unused-vars */
import * as infer from '../tern/lib/infer';
import * as tern from '../tern/lib/tern';
import * as walk from 'acorn-walk';
import * as fs from 'fs';
import defs from './aura_types.json';

const WG_DEFAULT_EXPORT = 95;
let server: any = {};

let shouldFilter = false;
/* this is necessary to inform the parameter types of the controller when
    the helper method is deleted */
const ForAllProps_Purgeable = infer.constraint({
    construct: function (c) {
        this.c = c;
    },
    addType: function (type) {
        if (!(type instanceof infer.Obj)) {
            return;
        }
        type.forAllProps(this.c);
    },
    purge: function (test) {
        if (this.sources) {
            for (let i = 0; i < this.sources.length; i++) {
                this.sources[i].purge(test);
            }
        }
    },
    addSource: function (source) {
        if (!this.sources) {
            this.sources = [];
        }
        this.sources.push(source);
    },
});

const getFilename = (filename): string => {
    if (server.options.projectDir.endsWith('/')) {
        return server.options.projectDir + filename;
    }
    return server.options.projectDir + '/' + filename;
};

const isBlocklisted = (filename): any => {
    let ret = filename.endsWith('/scrollerLib/bootstrap.js');
    ret = ret || filename.endsWith('ExportSymbolsHelper.js');
    return ret;
};

const readFile = async (filename): Promise<string> => {
    let normalized = filename;
    if (!normalized.startsWith('/')) {
        normalized = getFilename(normalized);
    }

    if (isBlocklisted(normalized)) {
        return '';
    }

    try {
        return fs.readFileSync(normalized, 'utf-8');
    } catch (e) {
        if (e.code === 'ENOENT') {
            return '';
        }
        throw e;
    }
};

const baseName = (path): any => {
    const lastSlash = path.lastIndexOf('/');
    if (lastSlash === -1) {
        return path;
    } else {
        return path.slice(lastSlash + 1);
    }
};

const trimExt = (path): any => {
    const lastDot = path.lastIndexOf('.');
    if (lastDot === -1) {
        return path;
    } else {
        return path.slice(0, lastDot);
    }
};

const initScope = (scope): void => {
    const module = new infer.Obj();
    module.propagate(scope.defProp('module'));
    const exports = new infer.Obj(true);
    module.origin = exports.origin = scope.origin;
    module.originNode = exports.originNode = scope.originNode;
    exports.propagate(scope.defProp('exports'));
    const moduleExports = (scope.exports = module.defProp('exports'));
    exports.propagate(moduleExports, WG_DEFAULT_EXPORT);
};

const getName = (name, type): string => {
    const newname = name.replace(/Controller.js$|Helper.js$|Renderer.js$|Test.js$/, '') + type;
    return newname;
};

const getController = (name): string => getName(name, 'Controller.js');

const getHelper = (name): string => getName(name, 'Helper.js');

const getRenderer = (name): string => getName(name, 'Renderer.js');

const resolver = (file, parent): any => file;

const unloadDefs = (): void => {
    server.deleteDefs('Aura');
};

const readFileAsync = (filename, c): void => {
    readFile(filename).then(function (contents) {
        c(null, contents);
    });
};

const findAndBindComponent = (type, server, cx, infer): void => {
    const evs = cx.props['Component'];
    if (!evs) {
        return;
    }
    for (let z = 0; z < evs.length; z++) {
        const y = evs[z];
        if (y.name === 'Aura.Component') {
            const obj = y.props['Component'].types[0].props['prototype'].getObjType();
            const int = infer.getInstance(obj);
            int.propagate(type);
        }
    }
};

const findAndBindHelper = (type, server, modules, file): void => {
    const helperFile = getHelper(file.name);

    const bn = trimExt(baseName(helperFile));
    const r = server.findFile(helperFile);
    if (!r) {
        server.addFile(helperFile);
    }
    const helper = modules.resolveModule(helperFile);
    //  console.log("Resolved module" + helperFile);
    //  console.dir(helper);
    const hp = helper.getProp(bn);
    if (!hp.getType()) {
        // this handles new props added to the helper...
        helper.on('addType', function (helperType, val) {
            const p = new ForAllProps_Purgeable(function (prop, val, local) {
                if (bn === prop) {
                    val.propagate(type);
                }
            });
            p.addSource(type);
            helperType.propagate(p);
        });
    } else {
        // now we need to handle there were changes to the .cmp,
        // but not the helper,
        const p = new ForAllProps_Purgeable(function (prop, val, local) {
            if (bn === prop) {
                val.propagate(type);
            }
        });
        p.addSource(type);
        helper.propagate(p);
    }
};

const findAndBindEvent = (type, server, cx, infer): void => {
    // this is slightly hacky, but have no idea how to get the event Otherwise
    const evs = cx.props['Event'];
    if (!evs) {
        return;
    }
    for (let z = 0; z < evs.length; z++) {
        const y = evs[z];
        if (y.name === 'Aura.Event') {
            const obj = y.props['Event'].types[0].props['prototype'].getObjType();
            const int = infer.getInstance(obj);
            int.propagate(type);
        }
    }
};

const ternError = (msg): Error => {
    const err = new Error(msg);
    err.name = 'TernError';
    return err;
};

const connectModule = async (file, out): Promise<void> => {
    if (isBlocklisted(file.name)) {
        return;
    }

    server.startAsyncAction();
    const modules = infer.cx().parent.mod.modules;
    const cx = infer.cx();
    console.log('Starting... ' + file.name);
    if (/Helper.js$/.test(file.name)) {
        // need to reestablish server context after awaits
        infer.withContext(server.cx, function () {
            console.log('Process helper exports ' + file.name);
            let outObj;
            if (!out.getType()) {
                const type = baseName(file.name).replace(/.js$/, '');
                outObj = new infer.Obj(true);
                outObj.origin = file.name;
                outObj.originNode = file.ast;
                outObj.name = type;
                out.addType(outObj);
            } else {
                outObj = out.getType();
            }
            try {
                walk.simple(
                    file.ast,
                    {
                        ObjectExpression: function (node, state) {
                            const parent = infer.parentNode(node, file.ast);
                            const grand = infer.parentNode(parent, file.ast);
                            if (grand.type === 'Program') {
                                // add some jsdoc
                                // @ts-ignore - objType is a custom tern property
                                if (node.objType) {
                                    // @ts-ignore - objType is a custom tern property
                                    node.objType.doc =
                                        'A helper resource contains functions that can be reused by your JavaScript code in the component bundle. ';
                                }
                                //  node.objType.forAllProps( function(prop, val, local) {
                                //    val.propagate(outObj.defProp(prop));
                                //    });
                                // -- would have worked, but didnt'
                                // delete all types, and re-add...
                                const target = outObj.defProp(baseName(file.name).replace(/.js$/, ''));
                                const types = target.types;
                                while (types.length) {
                                    types.pop();
                                }
                                //note: propogate calls addType on the target
                                // todo: this could be made more efficient with a custom propogation strategy
                                // similar to ForAllProps_Purgeable
                                // @ts-ignore - objType is a custom tern property
                                if (node.objType) {
                                    try {
                                        // @ts-ignore - objType is a custom tern property
                                        node.objType.propagate(target);
                                    } catch (err) {
                                        console.error(err);
                                    }
                                }
                                //outObj.defProp(baseName(file.name).replace(/.js$/, ''))
                                // eslint-disable-next-line no-throw-literal
                                throw 'stop';
                            }
                        },
                    },
                    infer.searchVisitor,
                );
            } catch (stop) {
                if (stop !== 'stop') {
                    console.error(stop);
                    throw stop;
                }
            }
        });
        // We should also make sure that the controller is all up to date too...
        const controller = getController(file.name);
        try {
            const text = await readFile(controller);
            const sfile = server.findFile(controller);
            if (!sfile || sfile.text !== text) {
                server.addFile(controller, text);
            }
        } catch (ignore) {}
        const renderer = getRenderer(file.name);
        try {
            const text = await readFile(renderer);
            const sfile = server.findFile(renderer);
            if (!sfile || sfile.text !== text) {
                server.addFile(renderer, text);
            }
        } catch (ignore) {}
    }
    // reestablish scope after awaits
    infer.withContext(server.cx, function () {
        console.log('Fixing scopes...' + file.name);
        walk.simple(file.ast, {
            ObjectExpression: function (node, state) {
                const parent = infer.parentNode(node, file.ast);
                const grand = infer.parentNode(parent, file.ast);
                if (grand.type === 'Program') {
                    for (let i = 0; i < node.properties.length; ++i) {
                        // @ts-ignore - Property type narrowing
                        if (node.properties[i].value.type === 'FunctionExpression') {
                            // @ts-ignore - Property type narrowing
                            const val = node.properties[i].value;
                            const fn = val && val.scope && val.scope.fnType;
                            if (!fn || !fn.name) {
                                continue;
                            }

                            if (/Renderer.js$/.test(file.name)) {
                                //step 2, assign exported type to params
                                const cmp = fn.args[0];
                                const hlp = fn.args[1];
                                if (cmp) {
                                    findAndBindComponent(cmp, server, cx, infer);
                                }
                                if (hlp) {
                                    findAndBindHelper(hlp, server, modules, file);
                                }
                            } else if (/Helper.js$/.test(file.name)) {
                                //step 2, assign exported type to params
                                const cmp = fn.args[0];
                                if (cmp) {
                                    findAndBindComponent(cmp, server, cx, infer);
                                }
                            } else if (/Controller.js$/.test(file.name)) {
                                //step 2, assign exported type to params
                                const cmp = fn.args[0];
                                const evt = fn.args[1];
                                const hlp = fn.args[2];
                                if (evt) {
                                    findAndBindEvent(evt, server, cx, infer);
                                }
                                if (cmp) {
                                    findAndBindComponent(cmp, server, cx, infer);
                                }
                                if (hlp) {
                                    findAndBindHelper(hlp, server, modules, file);
                                }
                            }
                        }
                    }
                }
            },
        });
        console.log('All done ' + file.name);
    });

    server.finishAsyncAction();
};

tern.registerPlugin('aura', function (s, options) {
    server = s;
    if (!server.options.async) {
        throw Error('Server must be async');
    }
    server.options.getFile = readFileAsync;
    server.loadPlugin('modules');
    server.mod.modules.on('wrapScope', initScope);
    server.mod.modules.on('getExports', connectModule);
    server.mod.modules.resolvers.push(resolver);
    let currentQuery;

    server.on('completion', function (file, query) {
        // don't hijack the request to retrieve the standard completions
        if (currentQuery === query) {
            return;
        }
        currentQuery = query;
        // request the standard completions
        let filteredResult;
        query.docFormat = 'full';
        server.request(
            {
                query: query,
            },
            function (err, result) {
                if (err) {
                    console.log(err);
                }
                if (shouldFilter) {
                    result.completions = result.completions.filter(function (completion, index, array) {
                        const accepted =
                            (completion.doc && completion.doc.indexOf('@platform') !== -1 && completion.origin === 'Aura') || completion.origin !== 'Aura';
                        if (accepted && completion.doc) {
                            completion.doc = completion.doc.split('\n@')[0];
                            completion.doc = completion.doc.replace('@description', '');
                        }
                        return accepted;
                    });
                }
                filteredResult = result;
                // reset for future queries
                currentQuery = undefined;
            },
        );
        return filteredResult;
    });

    console.log('IDE mode');
    server.addDefs(defs);

    console.log(new Date().toISOString() + ' Done loading!');
});

tern.defineQueryType('ideInit', {
    run: function (server, query) {
        if (query.unloadDefs) {
            unloadDefs();
            console.log('Unloaded default Aura defs');
        }

        if (query.shouldFilter === true || query.shouldFilter === false) {
            shouldFilter = query.shouldFilter;
        }
        return 'OK';
    },
});

tern.defineQueryType('cleanup-file', {
    run: function (server, query) {
        const files = query.files;
        files.forEach(function (f) {
            let ff = f;
            if (ff.startsWith('/')) {
                ff = ff.slice(1);
            }
            const m = server.mod.modules.modules[ff];
            if (m) {
                m.purge(function (type) {
                    if (type instanceof ForAllProps_Purgeable) {
                        return false;
                    } else {
                        if (type.origin && type.origin === ff) {
                            return true;
                        }
                    }
                    return false;
                });
                delete server.mod.modules.modules[ff];
            }
        });
        return 'OK';
    },
});

tern.defineQueryType('guess-types', {
    takesFile: true,
    run: function (server, query, file) {
        if (!query.end) {
            throw ternError('missing .query.end field');
        }

        if (!query.property) {
            throw ternError('missing .query.property field');
        }

        const start = tern.resolvePos(file, query.end);
        const types = [];

        const gather = (prop, obj, depth): void => {
            const val = obj.props[prop];
            const type = infer.toString(val.getType());
            types.push({
                property: prop,
                type: type,
                // The following causes J2V8 to crash
                //parent: obj.getType(),
                depth: depth,
            });
        };

        infer.forAllLocalsAt(file.ast, start, file.scope, gather);
        return {
            locals: types,
        };
    },
});

tern.defineQueryType('reset', {
    takesFile: false,
    run: function (server, query, file) {
        server.reset();
        return 'OK';
    },
});
