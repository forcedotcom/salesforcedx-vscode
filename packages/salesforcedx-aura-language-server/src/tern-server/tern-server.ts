import fs, { readFileSync } from 'fs';
import * as tern from '../tern/lib/tern';
import path from 'path';
import * as util from 'util';
import * as infer from '../tern/lib/infer';
import LineColumnFinder from 'line-column';
import { findPreviousWord, findPreviousLeftParan, countPreviousCommas } from './string-util';
import URI from 'vscode-uri';
import browser from '../tern/defs/browser.json';
import ecmascript from '../tern/defs/ecmascript.json';

import { memoize } from '@salesforce/salesforcedx-lightning-lsp-common';
import {
    TextDocumentPositionParams,
    CompletionList,
    CompletionItem,
    Hover,
    Location,
    TextDocumentChangeEvent,
    CompletionParams,
    Position,
    Range,
    ReferenceParams,
    SignatureHelp,
    SignatureInformation,
    Definition,
} from 'vscode-languageserver';

interface TernServer extends tern.Server {
    files: TernFile[];
    cx: any;
    normalizeFilename(file: string): string;
    /**
     * Register a file with the server. Note that files can also be included in requests. When using this
     * to automatically load a dependency, specify the name of the file (as Tern knows it) as the third
     * argument. That way, the file is counted towards the dependency budget of the root of its dependency graph.
     */
    addFile(name: string, text?: string, parent?: string): void;
    /** Unregister a file. */
    delFile(name: string): void;
    /** Forces all files to be fetched an analyzed, and then calls the callback function. */
    flush(callback: () => void): void;
    /**
     * Perform a request. `doc` is a (parsed) JSON document as described in the protocol documentation.
     * The `callback` function will be called when the request completes. If an `error` occurred,
     * it will be passed as a first argument. Otherwise, the `response` (parsed) JSON object will be passed as second argument.
     *
     * When the server hasn't been configured to be asynchronous, the callback will be called before request returns.
     */
    request(doc: any, callback: any): void;
}
interface TernFile {
    name: string;
    text: string;
}

let theRootPath: string;
let ternServer: TernServer;
let asyncTernRequest;
let asyncFlush;

const defaultConfig = {
    ecmaVersion: 6,
    stripCRs: false,
    disableLoadingLocal: true,
    verbose: true,
    debug: true,
    async: true,
    dependencyBudget: 20000,
};

const auraInstanceLastSort = (a: string, b: string): number =>
    a.endsWith('AuraInstance.js') === b.endsWith('AuraInstance.js') ? 0 : a.endsWith('AuraInstance.js') ? 1 : -1;

const loadPlugins = async (): Promise<{ aura: true; modules: true; doc_comment: true }> => {
    await import('./tern-aura');
    await import('../tern/plugin/modules');
    await import('../tern/plugin/doc_comment');

    return {
        aura: true,
        modules: true,
        doc_comment: true,
    };
};

/** recursively search upward from the starting diretory.  Handling the is it a monorepo vs. packaged vs. bundled code  */
const searchAuraResourcesPath = (dir: string): string => {
    console.log(`aura-language-server: searching for resources/aura in ${dir}`);
    if (fs.existsSync(path.join(dir, 'resources', 'aura'))) {
        console.log('found resources/aura in', dir);
        return path.join(dir, 'resources', 'aura');
    }
    if (path.dirname(dir) === dir) {
        throw new Error('No resources/aura directory found');
    }
    return searchAuraResourcesPath(path.dirname(dir));
};

const ternInit = async (): Promise<void> => {
    await asyncTernRequest({
        query: {
            type: 'ideInit',
            unloadDefs: true,
            // shouldFilter: true,
        },
    });
    const resources = searchAuraResourcesPath(__dirname);
    (await fs.promises.readdir(resources, { withFileTypes: true, recursive: true }))
        .filter((dirent) => dirent.isFile() && dirent.name.endsWith('.js'))
        .map((dirent) => path.join(dirent.parentPath, dirent.name))
        // special handling for hacking one snowflake file that needs to go last
        .sort(auraInstanceLastSort)
        .map((file) => ({
            file,
            contents: file.endsWith('AuraInstance.js')
                ? // and the snowflake needs to me modified
                  readFileSync(file, 'utf-8').concat(`\nwindow['$A'] = new AuraInstance();\n`)
                : readFileSync(file, 'utf-8'),
        }))
        .map(({ file, contents }) => ternServer.addFile(file, contents));
};

const init = memoize(ternInit);

export const startServer = async (rootPath: string, wsroot: string): Promise<tern.Server> => {
    const defs = [browser, ecmascript];
    const plugins = await loadPlugins();
    const config: tern.ConstructorOptions = {
        ...defaultConfig,
        defs,
        plugins,
        projectDir: rootPath,
        getFile(filename: string, callback: (error: Error | undefined, content?: string) => void): void {
            // note: this isn't invoked
            fs.readFile(path.resolve(rootPath, filename), 'utf8', callback);
        },
    };
    theRootPath = wsroot;
    ternServer = new tern.Server(config) as TernServer;
    asyncTernRequest = util.promisify(ternServer.request.bind(ternServer));
    asyncFlush = util.promisify(ternServer.flush.bind(ternServer));

    init();

    return ternServer;
};

const lsp2ternPos = ({ line, character }: { line: number; character: number }): tern.Position => ({ line, ch: character });

const tern2lspPos = ({ line, ch }: { line: number; ch: number }): Position => ({ line, character: ch });

const fileToUri = (file: string): string => {
    if (path.isAbsolute(file)) {
        return URI.file(file).toString();
    } else {
        return URI.file(path.join(theRootPath, file)).toString();
    }
};

const uriToFile = (uri: string): string => {
    const parsedUri = URI.parse(uri);
    // paths from tests can be relative or absolute
    return parsedUri.scheme ? parsedUri.fsPath : uri;
};

const tern2lspRange = ({ start, end }: { start: tern.Position; end: tern.Position }): Range => ({
    start: tern2lspPos(start),
    end: tern2lspPos(end),
});

const tern2lspLocation = ({ file, start, end }: { file: string; start: tern.Position; end: tern.Position }): Location => ({
    uri: fileToUri(file),
    range: tern2lspRange({ start, end }),
});

const ternRequest = async (event: TextDocumentPositionParams, type: string, options: any = {}): Promise<any> =>
    await asyncTernRequest({
        query: {
            type,
            file: uriToFile(event.textDocument.uri),
            end: lsp2ternPos(event.position),
            lineCharPositions: true,
            ...options,
        },
    });

export const addFile = (event: TextDocumentChangeEvent): void => {
    const { document } = event;
    ternServer.addFile(uriToFile(document.uri), document.getText());
};

export const delFile = (close: TextDocumentChangeEvent): void => {
    const { document } = close;
    ternServer.delFile(uriToFile(document.uri));
};

export const onCompletion = async (completionParams: CompletionParams): Promise<CompletionList> => {
    try {
        await init();
        await asyncFlush();

        const { completions } = await ternRequest(completionParams, 'completions', {
            types: true,
            docs: true,
            depths: true,
            guess: true,
            origins: true,
            urls: true,
            expandWordForward: true,
            caseInsensitive: true,
        });
        const items: CompletionItem[] = completions.map((completion) => {
            let kind = 18;
            if (completion.type && completion.type.startsWith('fn')) {
                kind = 3;
            }
            return {
                documentation: completion.doc,
                detail: completion.type,
                label: completion.name,
                kind,
            };
        });
        return {
            isIncomplete: true,
            items,
        };
    } catch (e) {
        if (e.message && e.message.startsWith('No type found')) {
            return;
        }
        return {
            isIncomplete: true,
            items: [],
        };
    }
};

export const onHover = async (textDocumentPosition: TextDocumentPositionParams): Promise<Hover> => {
    try {
        await init();
        await asyncFlush();
        const info = await ternRequest(textDocumentPosition, 'type');

        const out = [];
        out.push(`${info.exprName || info.name}: ${info.type}`);
        if (info.doc) {
            out.push(info.doc);
        }
        if (info.url) {
            out.push(info.url);
        }

        return { contents: out };
    } catch (e) {
        if (e.message && e.message.startsWith('No type found')) {
            return;
        }
    }
};

export const onTypeDefinition = async (textDocumentPosition: TextDocumentPositionParams): Promise<Definition> => {
    const info = await ternRequest(textDocumentPosition, 'type');
    if (info && info.origin) {
        const contents = fs.readFileSync(info.origin, 'utf-8');
        const endCol = new LineColumnFinder(contents, { origin: 0 }).fromIndex(contents.length - 1);
        return {
            uri: fileToUri(info.origin),
            range: {
                start: {
                    line: 0,
                    character: 0,
                },
                end: {
                    line: endCol.line,
                    character: endCol.col,
                },
            },
        };
    }
};

export const onDefinition = async (textDocumentPosition: TextDocumentPositionParams): Promise<Location> => {
    try {
        await init();
        await asyncFlush();
        const { file, start, end } = await ternRequest(textDocumentPosition, 'definition', { preferFunction: false, doc: false });
        if (file) {
            const responseURI = fileToUri(file);
            // check to see if the request position is inside the response object
            const requestURI = textDocumentPosition.textDocument.uri;
            if (
                responseURI === requestURI &&
                start.line === textDocumentPosition.position.line &&
                textDocumentPosition.position.character >= start.ch &&
                textDocumentPosition.position.character <= end.ch
            ) {
                return onTypeDefinition(textDocumentPosition) as any;
            }
            if (file === 'Aura') {
                return;
            } else if (file.indexOf('/resources/aura/') >= 0) {
                const slice = file.slice(file.indexOf('/resources/aura/'));
                const real = path.join(__dirname, '..', '..', slice);
                return {
                    uri: URI.file(real).toString(),
                    range: tern2lspRange({ start, end }),
                };
            }
            return tern2lspLocation({ file, start, end });
        }
    } catch (e) {
        if (e.message && e.message.startsWith('No type found')) {
            return;
        }
    }
};

export const onReferences = async (reference: ReferenceParams): Promise<Location[]> => {
    await init();
    await asyncFlush();
    const { refs } = await ternRequest(reference, 'refs');
    if (refs && refs.length > 0) {
        return refs.map((ref) => tern2lspLocation(ref));
    }
};

export const onSignatureHelp = async (signatureParams: TextDocumentPositionParams): Promise<SignatureHelp> => {
    const {
        position,
        textDocument: { uri },
    } = signatureParams;
    try {
        await init();
        await asyncFlush();
        const files = ternServer.files;
        const fileName = ternServer.normalizeFilename(uriToFile(uri));
        const file = files.find((f) => f.name === fileName);

        const contents = file.text;

        const offset = new LineColumnFinder(contents, { origin: 0 }).toIndex(position.line, position.character);

        const left = findPreviousLeftParan(contents, offset - 1);
        const word = findPreviousWord(contents, left);

        const info = await asyncTernRequest({
            query: {
                type: 'type',
                file: file.name,
                end: word.start,
                docs: true,
            },
        });

        const commas = countPreviousCommas(contents, offset - 1);
        const cx = ternServer.cx;
        let parsed;
        infer.withContext(cx, () => {
            const parser = new infer.def.TypeParser(info.type);
            parsed = parser.parseType(true);
        });

        const params = parsed.args.map((arg, index) => {
            const type = arg.getType();
            return {
                label: parsed.argNames[index],
                documentation: type.toString() + '\n' + (type.doc || ''),
            };
        });

        const sig: SignatureInformation = {
            label: parsed.argNames[commas] || 'unknown param',
            documentation: `${info.exprName || info.name}: ${info.doc}`,
            parameters: params,
        };
        const sigs: SignatureHelp = {
            signatures: [sig],
            activeSignature: 0,
            activeParameter: commas,
        };
        return sigs;
    } catch (e) {
        // ignore
    }
};
