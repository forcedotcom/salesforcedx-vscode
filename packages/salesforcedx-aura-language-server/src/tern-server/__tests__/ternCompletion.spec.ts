import { AuraWorkspaceContext } from '../../context/aura-context';
import { startServer, onCompletion, onHover, onDefinition, onReferences } from '../tern-server';

it('tern completions', async () => {
    const ws = 'test-workspaces/sfdx-workspace';
    const context = new AuraWorkspaceContext(ws);
    await context.configureProject();

    await startServer(ws, ws);
    const completions = await onCompletion({
        textDocument: {
            uri: 'force-app/main/default/aura/lightningExamplesApp/lightningExamplesAppController.js',
        },
        position: {
            line: 0,
            character: 0,
        },
    });
    expect(completions).toMatchSnapshot();
});

it('tern hover', async () => {
    const hover = await onHover({
        textDocument: {
            uri: 'force-app/main/default/aura/lightningExamplesApp/lightningExamplesAppController.js',
        },
        position: {
            line: 2,
            character: 10,
        },
    });
    expect(hover).toMatchSnapshot();
});

it('tern definition, same file', async () => {
    const helper = await onDefinition({
        textDocument: {
            uri: 'force-app/main/default/aura/lightningExamplesApp/lightningExamplesAppController.js',
        },
        position: {
            line: 2,
            character: 10,
        },
    });
    expect(helper).toMatchSnapshot();
});

it('tern references', async () => {
    const functionInsideHelper = await onReferences({
        textDocument: {
            uri: 'force-app/main/default/aura/lightningExamplesApp/lightningExamplesAppHelper.js',
        },
        position: {
            line: 1,
            character: 11,
        },
        context: {
            includeDeclaration: false,
        },
    });
    expect(functionInsideHelper).toMatchSnapshot();
});
