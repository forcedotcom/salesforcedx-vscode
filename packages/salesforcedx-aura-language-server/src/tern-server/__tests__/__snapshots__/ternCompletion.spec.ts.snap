// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tern completion tern completions 1`] = `
{
  "isIncomplete": true,
  "items": [],
}
`;

exports[`tern completion tern definition, same file 1`] = `undefined`;

exports[`tern completion tern hover 1`] = `
{
  "contents": [],
}
`;

exports[`tern completion tern references 1`] = `
[
  {
    "range": {
      "end": {
        "character": 27,
        "line": 2,
      },
      "start": {
        "character": 15,
        "line": 2,
      },
    },
    "uri": "file:///Users/madhur.shrivastava/salesforcedx-vscode/test-workspaces/sfdx-workspace/force-app/main/default/aura/lightningExamplesApp/lightningExamplesAppController.js",
  },
  {
    "range": {
      "end": {
        "character": 16,
        "line": 1,
      },
      "start": {
        "character": 4,
        "line": 1,
      },
    },
    "uri": "file:///Users/madhur.shrivastava/salesforcedx-vscode/test-workspaces/sfdx-workspace/force-app/main/default/aura/lightningExamplesApp/lightningExamplesAppHelper.js",
  },
]
`;
